### ISSUE TITLE
Http rest API

### ISSUE THREAD

**Issue by benjaminVadon** [mirage/irmin#415]
Since 1.0.0 version http high-level bindings were removed because of complexity in previous version and uncertain use by anybody, as explained here #414 by @samoht .

This issue is aimed to propose a new basic Rest API for Irmin. This way, using Irmin as a standalone executable, this opens it to a wider horizon than just Ocaml world.
   
Reading code of new http modules, it's seems easier than before to build such rest API, and @samoht proposed to help me to start this new backend.

**Comment by dudelson** [comment]
I am interested in fleshing this out into a GSoC project and would like to understand the scope. Right now Irmin provides an http server with a REST API of sorts, but it only supports [endpoints](https://github.com/mirage/irmin/blob/master/src-http/irmin_http_server.ml#L310) like /blobs, /trees, /commits, or specific IDs. Formerly, there were endpoints like /graph, but these were removed recently for the reasons above. We want to add these back, as well as endpoints like clone, read, write, etc. that allow users to interact with the store. Is that right?

**Comment by samoht** [comment]
Yes that's right. The HTTP API in 1.0 exposes only endpoints for low-level resources (blogs, tree, commits, tags). From this, a "clever" HTTP client can reconstruct the full Irmin API, by doing most of the computation on the client side (e.g. this is done [here](https://github.com/mirage/irmin/blob/master/src-http/irmin_http.ml#L439)).

For client in different langages, this is not practical as it means re-implementing most of the Irmin core logic. The proper way forward is thus to do most of the computions on the server side, expose a richer/higher-level API than the current low-level one and have only a very thin client accessing that high level API.

So I would recommend to:
1. design the new API in term of high-level resources. We probably don't want to expose the full [API](http://mirage.github.io/irmin/Irmin.S.html) but it would be great to expose as much of it as possible. 
2. the hardest issue is probably handling of transactions: do we want to handle transaction client or server side? We should probably start by doing that client side, but in the longer term a proper API would probably expose both options.
3. I quite like [webmachine](https://github.com/inhabitedtype/ocaml-webmachine), which is already used by the low-level API. It would be great if the high-level API would use it as well. 
4. a proper high-level API should also expose some sort of authentication (even very basic). Webmachine could help.

Please ask any questions if you are blocked, either here or on the mailing list. /cc @talex5 who knows very well irmin and ocaml-webmachine and @seliopou who create ocaml-webmachine in case they want to step in.

**Comment by talex5** [comment]
Another problem will likely be implementing watches. e.g. it's hard to get predictable behaviour if you allow the client to open multiple TCP connections to the server.

**Comment by samoht** [comment]
There is already some code to dispatch multiple watches using one connection (as we already have to do this when watching for filesystem changes). So the watch functions in the low-level HTTP implementation are supposed to do "the right thing" (and there are some tests to check this). Obviously it would be great to check that's indeed in the case :-) And this should remain true to the high-level API too.

**Comment by dudelson** [comment]
I'm in the process of expanding on the technical section of my proposal and I have some clarifying questions:

- I'm a bit confused as to the types of endpoints we want to expose (for the purposes of forming "expected outcomes" and "reach goals" for GSoC). I was thinking that the high-level REST API should be very similar to the current CLI; e.g. instead of functions exposed in the Ocaml API like `mem_tree`, `find`, `set`, and `merge`, the REST API would expose things like `/read`, `/write`, `/snapshot`, and `/revert`. Clearly these are two very different APIs, so I want to make sure that the one I have in mind is the one that makes the most sense.
- I don't understand the issue of handling transactions. I was thinking that a transaction would be an implicit part of some of the endpoints, e.g. POSTing to the `/write` endpoint would mean the server would initiate a transaction and the result would be sent back to the client. What is the reason for wanting to enable the client to handle transactions, and how does this process work? It seems like having the client handle transactions requires duplication of Irmin code on the client-side, i.e. the client isn't so "thin" anymore.
- I also don't understand why it will likely be difficult to get predictable behaviour when a client is setting multiple watches simultaneously via multiple open TCP streams to the same server. Since the watch and transaction APIs appear to be thread-safe, in what scenario does predictable behaviour become a problem?

Additionally, since I am proposing a new idea not found on [canopy](http://canopy.mirage.io/), I wanted to confirm that should my proposal be accepted, @samoht would be my mentor for the summer.

**Comment by talex5** [comment]
Transactions: Often clients need to update several files at once, such that either they're all updated or none of them are, and no other client sees an inconsistent state.

Watches: For example, a typical client pattern is:

1. Start watching for changes
2. Read the current value

This ensures that the client will always eventually see the most up-to-date value, because any change after (2) will be reported to it. If (1) and (2) are on different HTTP connections then the messages may arrive out of order, and this may happen:

1. The request for the current value arrives and the current value is returned.
2. The value is updated on the server.
3. The request for a watch arrives; further changes will be reported.

In this case update (2) is never reported to the client. There are ways arround this (e.g. sending the last-known hash when setting up the watch).

**Comment by dudelson** [comment]
Transactions: I'm not 100% clear on how the block store structure looks internally. Is it not possible to update several different keys in the same commit? If not, then by having clients deal with transactions, we would be implementing something like [this](https://stackoverflow.com/questions/9056009/restful-atomic-update-of-multiple-resources#10044965)?

Watches: I see how that works now. Couldn't we also collapse the client pattern into one step by returning the current value as part of the acknowledgement that the watch request was received? That way, there's only one communication that has to be sent, and the client still receives the most up-to-date value after requresting a watch.

**Comment by samoht** [comment]
To try to answer to your questions @dudelson

- About the scope of the API: people seem to want to use the HTTP API to ease making bindings with other languages. A good goal for this new HTTP API would probably be to have at least a client written in an other language (pick the one you are the more familiar too) and show that you can do something useful with it. The CLI API was not really designed fully, so I am not sure it is the right model, but I don't see any issue of having something a bit higher-level than the OCaml API (and it certainly doesn't have to be as complete as the OCaml API).

- As @talex5 said, transactions allow to perform multiple writes and reads, on a long time span. Existing client/server protocols which use transactions usually defines `start-transaction` and `end-transaction` messages as well as prefix all existing messages with a transaction ID. But then, servers have to remember which transaction are opened, and have to dispatch the client messages to the right open transaction: this adds a layer of complexity that we maybe do not want (need to evaluate the pros. and cons.). An intermediate solution would be to allow atomic updates of a `tree` object (e.g. to allow atomic multiple writes).

**Comment by samoht** [comment]
Yes, the link that you gave is indeed a possible implementation of long-running transactions. The cons with that approach is that 1/ everyone can add to any transaction (e.g. there is some security implications) and 2/ that the server needs to keep track of open transaction (and probably want to auto-close old transactions to save resources).

Also:

> Additionally, since I am proposing a new idea not found on canopy, I wanted to confirm that should my proposal be accepted, @samoht would be my mentor for the summer.

I confirm that @yomimono and I will be available to mentor you over the summer.

**Comment by dudelson** [comment]
@samoht I have submitted a proposal draft through the GSoC system and also sent a copy to the mailing list; review would be appreciated. Feedback from others familiar with Irmin is also great!

**Comment by samoht** [comment]
Thanks! The draft looks very good and I have added some notes on the doc. I never mentored for GSoC before, so I hope my remarks make sense.

**Comment by benjaminVadon** [comment]
@dudelson or @samoht can you provide a link to this GSoC draft please ?

**Comment by dudelson** [comment]
Yep! You can view it [here](https://docs.google.com/document/d/1Vu97h2gNuJMTEYdIcghFfo_9L4GtCpEprzWfOivhy6g/edit?usp=sharing).

**Comment by dudelson** [comment]
I dug into the code a bit today and have some questions. Because they're not directly related to the REST API, I sort of hesitate to ask them here, but in the end I couldn't think of a better place to ask (most of the questions seem too specific to use the mailing list for). If there's a better place for me to inquire in the future, please let me know.

- There are places ([here](https://github.com/mirage/irmin/blob/master/src/irmin/conf.mli#L20), [here](https://github.com/mirage/irmin/blob/master/src/irmin/irmin.mli#L433)) in the codebase where a single polymorphic variant is used in a type definition. Why is `['Msg of string]` preferable to simply `string`?
- Does this project have coding conventions? I couldn't find any documentation related to this. Things like tab width and brace styles I can infer from the source, but I am confused about less obvious things, like why some module names are in all caps. What does this signify?
- From all of the reading about Irmin that I've done so far, I can't tell what the original motivation for starting the project was. Obviously by this point it has great utility as a database both inside and outside Mirage OS, but was it originally intended to fill Mirage's need for a pure-ocaml database, or was it borne out of theoretical research on mergeable functional data structures and the like?

**Comment by samoht** [comment]
Hi,

All of these questions have their place on the mailing list :-) but I am fine to answer these here:

- The OCaml community generally like to see types as a great way to document the semantics of functions. While `string` is very generic and doesn't carry any specific meaning `['Conflict of string]` carries a bit more of information for the users of that type. The `['Msg of string]` is another convention, set by the [Rresult](http://erratique.ch/software/rresult/doc/Rresult.R.html#msgs) library and usually followed by other MirageOS libs as well.

- No coding conventions, just try to follow https://ocaml.org/learn/tutorials/guidelines.html (but feel free to be flexible when this is needed :p). Module signatures (e.g. types of modules) are usually in caps, and module names are in caps when they are an abbreviations (e.g. `KV` for `Key`-`Value`, etc). I can try to write down some of the rules if you think that's useful.

- Initially, Irmin was written as a "new new" version of [Xenstored](http://gazagnaire.org/pub/GH09.pdf) which distribution and full audit/debug capabilities, e.g. see [the corresponding blog post](https://mirage.io/blog/introducing-irmin-in-xenstore) for a prototype of what we were aiming at. We though that this style of coordination layer, e.g. a distributed Git-like configuration database, would also be useful to coordinate MirageOS unikernels. Which leads to new way to handle conflicts (because Git is nice until you had to resolve conflicts). Which leads to mergeable data structures. Which leads to Irmin as it is today :-)

**Comment by dudelson** [comment]
Thanks for the explanation of coding conventions and the Irmin history lesson, they're helping me to understand the library better.

With regard to the explanation about polymorphic variants, I've read Daniel's excellent [stack overflow answer](https://stackoverflow.com/questions/9367181/variants-or-polymorphic-variants) about use cases for polymorphic variants, and it seems like the use case in [conf.mli](https://github.com/mirage/irmin/blob/master/src/irmin/conf.mli#L20) fits point number 4. That is, we want to attach some information about the purpose of the string, and the `'a parser` type is the only type that needs this information, so it's not worth creating a new type, rather we'd like to attach this information in a lightweight way. However, in the cases of the [Rresult library](http://erratique.ch/software/rresult/doc/Rresult.R.html#msgs) and [irmin.mli](https://github.com/mirage/irmin/blob/master/src/irmin/irmin.mli#L433), I still don't understand what the advantage of writing `type conflict = [ 'Conflict of string ]` over `type conflict = string` is. With the latter, don't you get the same semantic information in a less verbose, less error-prone way?

**Comment by dbuenzli** [comment]
> I still don't understand what the advantage of writing type conflict = [ 'Conflict of string ] over type conflict = string is. With the latter, don't you get the same semantic information in a less verbose, less error-prone way?

You can open such a polyvar (i.e. have your functions return `[> conflict]`)  which allows you to compose it with other error returning functions while keeping the source of the error. See [here](http://erratique.ch/software/rresult/doc/Rresult.html#usage).

**Comment by Tails** [comment]
What does this mean for the status of the HTTP API now? Have [all methods](https://github.com/dudelson/irmin/blob/6b9794ecae7c3d4b0b834b2a3a5bc807babf0166/src/irmin-http/swagger.yml) been removed? Because it is easier to POST a new JSON document than to embed it inside a GraphQL mutation query.
